apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/ClientBuilder.cs

```diff
@@ -0,0 +1,322 @@
+            clone.Endpoint = endpoint;
+            return clone;
+        }
+
+        /// <summary>
+        /// Creates a new builder based on this one, but with the specified credential scopes.
+        /// </summary>
+        /// <param name="scopes">The scopes to use when creating credentials, or null to use the default set of scopes.</param>
+        /// <returns>A builder with the specified scopes.</returns>
+        public TBuilder WithScopes(IEnumerable<string> scopes)
+        {
+            var clone = Clone();
+            clone.Scopes = scopes?.ToList().AsReadOnly();
+            return clone;
+        }
+
+        /// <summary>
+        /// Creates a new builder based on this one, but with the specified channel credentials.
+        /// </summary>
+        /// <remarks>
+        /// This method overrides any credentials specified with previous calls.
+        /// </remarks>
+        /// <param name="channelCredentials">The channel credentials to use, or null to use the default credentials.</param>
+        /// <returns>A builder with the specified channel credentials.</returns>
+        public TBuilder WithChannelCredentials(ChannelCredentials channelCredentials)
+        {
+            var clone = Clone();
+            clone.ClearCredentials();
+            clone.ChannelCredentials = channelCredentials;
+            return clone;
+        }
+
+        /// <summary>
+        /// Creates a new builder based on this one, but with the specified credential file.
+        /// </summary>
+        /// <remarks>
+        /// This method overrides any credentials specified with previous calls.
+        /// </remarks>
+        /// <param name="file">The credential file to use, or null to use the default credentials.</param>
+        /// <returns>A builder with the specified credential file.</returns>
+        public TBuilder WithCredentialsFile(string file)
+        {
+            // Note: don't load this until we build, so that any scope changes will be effective.
+            var clone = Clone();
+            clone.ClearCredentials();
+            clone.CredentialsFile = file;
+            return clone;
+        }
+
+        /// <summary>
+        /// Creates a new builder based on this one, but with the specified JSON credentials.
+        /// </summary>
+        /// <remarks>
+        /// This method overrides any credentials specified with previous calls.
+        /// </remarks>
+        /// <param name="json">The JSON credentials to use, or null to use the default credentials.</param>
+        /// <returns>A builder with the specified JSON credentials.</returns>
+        public TBuilder WithJsonCredentials(string json)
+        {
+            var clone = Clone();
+            clone.ClearCredentials();
+            clone.JsonCredentials = json;
+            return clone;
+        }
+
+        // TODO: Handle the fact that GoogleCredential implements ITokenAccess explicitly. That means the sample here doesn't work
```

Comment:
We can make whatever changes to the auth code to help with this.

Reply:
Great. I've had a couple of ideas, but we can explore those later.